File: Solutions/day1/py/main.py
def part_a(arr) -> int:
    count = 0
    prev = None
    for curr in arr:
        if prev is not None and prev < curr:
            count += 1
        prev = curr
    return count


def part_b(arr) -> int:
    prev = None
    count = 0
    for i in range(len(arr)):
        arr_slice = arr[i:i + 3]
        curr = sum(arr_slice)
        if len(arr_slice) == 3 and prev is not None and prev < curr:
            count += 1
        prev = curr
    return count


def main() -> tuple[int, int]:
    with open("../input.txt") as f:
        arr = [int(line) for line in f.read().split("\n") if line.isdigit()]
    return part_a(arr), part_b(arr)


if __name__ == '__main__':
    import time
    start_time = time.perf_counter()
    a, b = main()
    end_time = time.perf_counter()
    print(f"Result part_a: {a}")
    print(f"Result part_b: {b}")
    print(f"Time: {(end_time - start_time):4f} sec")

File: Solutions/day1/rust/src/main.rs
use std::fs;

fn part_a(v: &[i32]) -> i32{
    let mut prev: Option<i32> = None;
    let mut count: i32 = 0;
    for curr in v {
        if let Some(prev) = prev {
            if prev < *curr {
                count += 1;
            }
        }
        prev = Some(*curr)
    }
    count
}

fn part_b(v: &Vec<i32>) -> i32{
    let mut prev: Option<i32> = None;
    let mut count: i32 = 0;
    let mut curr: i32;
    for i in 0..v.len()-2 {
        curr = v[i] + v[i+1] + v[i+2];
        if let Some(prev) = prev {
            if prev < curr {
                count += 1;
            }
        }
        prev = Some(curr);
    }
    count
}

fn main() {
    let timer = std::time::Instant::now();
    let input_vec = fs::read_to_string("../input.txt")
        .unwrap()
        .lines()
        .map(|x| x.parse::<i32>().unwrap())
        .collect::<Vec<_>>();
    let a = part_a(&input_vec);
    let b = part_b(&input_vec);
    let total_time = timer.elapsed();
    println!("Result part_a: {}", a);
    println!("Result part_b: {}", b);
    println!("Time: {} sec", total_time.as_secs_f32());
}
